pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -695,25 +696,21 @@ func getProjectID(svc *compute.Service, projectNumberOrID string) (string, error
 }
 
 func getZonesForRegion(svc *compute.Service, projectID, region string) ([]string, error) {
-	// TODO: use PageToken to list all not just the first 500
-	listCall := svc.Zones.List(projectID)
-
-	// Filtering by region doesn't seem to work
-	// (tested in https://cloud.google.com/compute/docs/reference/latest/zones/list)
-	// listCall = listCall.Filter("region eq " + region)
-
-	res, err := listCall.Do()
-	if err != nil {
-		return nil, fmt.Errorf("unexpected response listing zones: %v", err)
-	}
 	zones := []string{}
-	for _, zone := range res.Items {
-		regionName := lastComponent(zone.Region)
-		if regionName == region {
+	filter := fmt.Sprintf("region eq %v", getRegionLink(svc, projectID, region))
```

Comment:
The zones change needs to be tested by creating a multi-zone cluster and dynamically provisioning gce pds explicitly in each zone.

Reply:
I'm new to k8s and don't yet fully understand how the TestGrid works, so apologies if this question sounds silly or totally unrelated, but isn't the [gce-multizone](https://k8s-testgrid.appspot.com/google-gce#gce-multizone) job already testing that? Otherwise, any pointer or direction would be appreciated, thanks!

Reply:
The multizone e2e tests currently do not include dynamic provisioning.  Also, they are run after the PR is submitted.  I think this change could have a big impact on multizone clusters, so it is better to test it ahead of time.

Reply:
This particular code does NOT change the behavior, it was already listing zones by region, the only change is that is using the `Filter` and `Pages` functions instead of the `Do` and a custom filter logic (and the new logic is extracted from the already existing [ListZonesInRegion](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce_zones.go#L72) function, which is used by the Service Load Balancer).

Anyway, I'd try to find how to add the dynamic provisioning tests, but it will take me some time until I learn how e2e tests work. Maybe I can remove this change from this PR (as it is not strictly necessary, there aren't' more than 3-4 zones per region), so we can move forward, and work on this part later?

Reply:
I know it shouldn't change the behavior, but there could always be some bug in the syntax or arguments, and the impact could completely break multizone clusters.  So I think it is important that is is thoroughly tested beforehand.

I am fine with handling this separately since the number of zones are not a lot in the first place.

Reply:
Understood, I'll update it to remove that change and will work on it later. Thanks for reviewing it!